### PR TITLE
fix issue 2968 by adding a new pragma(framework) 

### PR DESCRIPTION
**Pull request by benjones**
Which just adds appropriate flags to the linker command.

@jacob-carlborg I sort of based this off or your old patch.  LMK what I overlooked.

I copypasted stuff from pragma(lib) impl.  Not sure if it's a good idea to try to unify them or not.

I added this section to json.d.  I could only output it if it's not empty or something to reduce the number of tests that need to be updated?

### CODE REVIEW

File: test/runnable/test2968.d [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// DISABLED: win linux freebsd dragonflybsd netbsd
+
+struct __CFArray; //try to call some CF functions with strings
+alias CFArrayRef = __CFArray*;
+alias CFIndex = long;
+
+extern(Objective-C) CFArrayRef CFArrayCreate(void* allocator, const void** values, long numValues, void* cbs);
```
Comment (jacob-carlborg) [inline test/runnable/test2968.d]: These are not Objective-C functions. These are regular C functions.

File: test/runnable/test2968.d [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+extern(Objective-C) CFIndex CFArrayGetCount(CFArrayRef theArray);
+extern(Objective-C) const(void *) CFArrayGetValueAtIndex(CFArrayRef theArray, CFIndex idx);
+
+pragma(framework, "Foundation");
```
Comment (jacob-carlborg) [inline test/runnable/test2968.d]: * The above functions are part of the `CoreFoundation` framework

* I think the pragma should be at the top of the module

File: test/runnable/test2968.d [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+void main()
+{
+
+    ulong[5] array = [1,2,3,4,5];
+    auto cfa = CFArrayCreate(null, cast(void**)array.ptr, array.length, null);
+    const length = CFArrayGetCount(cfa);
+    assert(length == array.length);
+    foreach (i, x; array)
+    {
+        assert(x == cast(ulong) CFArrayGetValueAtIndex(cfa, i));
```
Comment (jacob-carlborg) [inline test/runnable/test2968.d]: You’re storing `int` not `ulong` on the array.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]
* Should this pragma only be available on macOS?

* I’m not sure if this is a good idea at all. GDC won’t support it since it doesn’t support `pragma(lib)`

**Comment by ibuclaw** [comment]
>     * I’m not sure if this is a good idea at all. GDC won’t support it since it doesn’t support `pragma(lib)`

Furthermore, even if GDC did support it.  There are still shortcomings to the pragma() approach that means it still won't work if you are linking to a project/package that has `pragma(framework)`.

File: test/runnable/test2968.d (hunk 3 above)
Comment (benjones) [reply test/runnable/test2968.d]: I am?  The array is declared as `ulong[5]` above

**Comment by benjones** [comment]
Thanks for the feedback.  If people decide that this isn't a good idea, then we should probably close the bug report wontfix.  

Is pragma(lib) useful in gdc?  It looks like only when the libs can be stored in the actual object file, but I don't know which formats actually support that.

@adamdruppe seemed enthusiastic about this before... do you still feel that way?

**Comment by adamdruppe** [comment]
Yes, I am still for it. Just because it doesn't work everywhere - even if it doesn't work most places - doesn't mean it isn't useful where it does work, and it still serves as formal, machine-readable documentation in the cases where it doesn't work.

File: test/runnable/test2968.d (hunk 3 above)
Comment (jacob-carlborg) [reply test/runnable/test2968.d]: Oh, I was only looking at the array literal. My bad.

**Comment by jacob-carlborg** [comment]
Needs a PR for the spec as well: https://dlang.org/spec/pragma.html.

**Comment by benjones** [comment]
Another note, since this writes the framework to the moduleDeps file, the framework info could be picked up by a build system tool

**Comment by kinke** [comment]
> Furthermore, even if GDC did support it, there are still shortcomings to the pragma() approach that means it still won't work if you are linking to a project/package that has pragma(framework).

+1. Extending the weak `pragma(lib)` (on Posix) approach would be a bad idea IMO. Linker directives make sense if they can be embedded in the object file (`pragma(lib)` and `pragma(linkerDirective)` on Windows), but can be easily misleading otherwise (need to import that .d source file in the compiler invocation performing the linking to have an effect). [And I'd find `framework` way too generic.]

**Comment by adamdruppe** [comment]
Just put a note in the documentation saying the result is implementation specific.

Don't let the perfect be the enemy of the good.

**Comment by Geod24** [comment]
I'm also concerned about this.
`pragma(lib)` sounded like a good idea, but in practice didn't pan out so well. `pragma(framework)`  limitations are already quite obvious. And willingly putting something that is implementation-defined is not solving the problem, it's ignoring it. And I don't see it enabling something that wasn't already possible, so IMO we should just mark this issue as WONTFIX.

**Comment by adamdruppe** [comment]
On Sun, Dec 01, 2019 at 06:30:59PM -0800, Mathias LANG wrote:
> `pragma(lib)` sounded like a good idea, but in practice didn't pan out so well. 

I find it enormously useful in practice and use it regularly without any
problem or difficulty.

**Comment by WalterBright** [comment]
I'll have to defer to those expert on the MacOS as to whether it is useful or not. Do other languages on the Mac support such a mechanism?

**Comment by jacob-carlborg** [comment]
> Do other languages on the Mac support such a mechanism?

Kind of. Objective-C and Swift supports autolinking of frameworks. Basically you only need to import the framework and it will automatically be linked when building.

**Comment by jacob-carlborg** [comment]
> Extending the weak pragma(lib) (on Posix) approach would be a bad idea IMO. Linker directives make sense if they can be embedded in the object file (pragma(lib) and pragma(linkerDirective) on Windows)

@kinke Linker directives can be embedded in Mach-O files, see: https://github.com/dlang/dmd/pull/10615#issuecomment-562126117

**Comment by kinke** [comment]
Well if it can in a suitable manner (embedded strings as additional linker cmdline arguments?), then I think it's a no-brainer to extend existing `pragma(linkerDirective)` to support these frameworks.

**Comment by kinke** [comment]
[Btw, LLD 9 added support for embedded library dependencies in ELF object files, suitable for `pragma(lib)`, see https://github.com/ldc-developers/ldc/issues/3245.]

**Comment by jacob-carlborg** [comment]
> Well if it can in a suitable manner (embedded strings as additional linker cmdline arguments?), then I think it's a no-brainer to extend existing pragma(linkerDirective) to support these frameworks.

As far as I can see it will embed strings as additional linker command line arguments. Here's an example of one of the load commands from the object dumper:

```
```

**Comment by kinke** [comment]
Okay perfect, that's analogous to MS COFF linker directives then. Then I'd propose either `pragma(linkerDirective, "-framework", "CoreFoundation")`, or `pragma(linkerDirective, "-framework CoreFoundation")` with the compiler tokenizing the string into possibly multiple linker flags.

**Comment by benjones** [comment]
Do you think there's an advantage to having the user write out the whole linker command rather than using the proposed syntax?  If we use linkerDirective, the user would have to static if it out on now macos builds, whereas pragma(framework) might only be examined on macos.  I'll defer to the experts, but I kind of like the pragma(framework) syntax.

**Comment by kinke** [comment]
Yes, I absolutely prefer having one generic way of expressing the same thing (linker flags) for multiple targets over a ever-growing family of target- and compiler-specific pragmas.

**Comment by adamdruppe** [comment]
I can live with linkerDirective instead of pragma framework if it actually works.

**Comment by benjones** [comment]
OK, I'll look into that approach.  Is there already code in the machobj backend for LC_LINKER_OPTION stuff or will that need to be added?

**Comment by jacob-carlborg** [comment]
> OK, I'll look into that approach. Is there already code in the machobj backend for LC_LINKER_OPTION stuff or will that need to be added?

No, that needs to be added. You can look at one of my pull requests how to add a new load command: https://github.com/dlang/dmd/pull/10476.

**Comment by ibuclaw** [comment]
> Well if it can in a suitable manner (embedded strings as additional linker cmdline arguments?), then I think it's a no-brainer to extend existing pragma(linkerDirective) to support these frameworks.

I had to do a double take, apparently I completely missed the introduction of that pragma.

**Comment by ibuclaw** [comment]
> [Btw, LLD 9 added support for embedded library dependencies in ELF object files, suitable for pragma(lib), see https://github.com/ldc-developers/ldc/issues/3245.]

I wrote such a thing for gdc too, which writes the data into a custom section, then all objects are scanned by the driver.

From what I remember, it didn't extend to linking against D libraries (neither static nor dynamic) that may have this section embedded in.  As the linked is patch for the LLVM linker, I assume you won't have that problem?

**Comment by kinke** [comment]
At least I don't plan to do any scanning of objects/libraries in the LDC driver; for Posix targets, we simply invoke gcc or clang for linking, so the plan is to let them take care of everything, even if that means that this special section might only work in combination with clang and LLD for now.

**Comment by kinke** [comment]
This is implemented for LDC now in https://github.com/ldc-developers/ldc/pull/3259. It includes a small frontend patch for `pragma(linkerDirective)` to
* be analyzed for all targets, not just MS COFF
* support multiple string arguments, not just 1

=> `pragma(linkerDirective, "-framework", "CoreFoundation");`

**Pull request closed without merging**